ENH: Optimize and cleanup ufunc calls and ufunc CheckOverrides

seberg opened the pull request:
This is a rather large commit which clean up the ufunc override
code for argument parsing. (I may attempt to split it up)
There are two main things, argument parsing, especially for
reductions is now much faster since `METH_FASTCALL` is used
(when keyword arguments are being used).
By moving the argument parsing and using more generic code this
also simplifies the ufunc override checking especially when
keyword arguments are present.
    
Both of this decreases the argument parsing overhead quite a lot
for ufunc calls, especially for small reductions. (Without double
checking, I believe the speedup was up to around 30% for small
reductions.)
    
The downside is some added/annoyance due to the use of marcos to
support both FASTCALL and no FASTCALL semantics.
As a side note: vectorcall is likely not a huge factor for ufuncs
since it is very common not to use any keyword arguments. OTOH,
code that uses ``out=...`` a lot on small arrays should see a nice
difference.

---

Only the last commit is new, the others are changes from gh-15269 and gh-15270
@mhvk the last commit here is about what you had once looked at for the override cleanup stuff.

mhvk commented:
Great!  Would think it is useful to split up. In particular, the benchmarks should go in first, to establish a baseline.

seberg committed:
ENH: Optimize and cleanup ufunc calls and ufunc CheckOverrides

This is a rather large commit which clean up the ufunc override
code for argument parsing. (I may attempt to split it up)
There are two main things, argument parsing, especially for
reductions is now much faster since `METH_FASTCALL` is used
(when keyword arguments are being used).
By moving the argument parsing and using more generic code this
also simplifies the ufunc override checking especially when
keyword arguments are present.

Both of this decreases the argument parsing overhead quite a lot
for ufunc calls, especially for small reductions. (Without double
checking, I believe the speedup was up to around 30% for small
reductions.)

The downside is some added/annoyance due to the use of marcos to
support both FASTCALL and no FASTCALL semantics.
As a side note: vectorcall is likely not a huge factor for ufuncs
since it is very common not to use any keyword arguments. OTOH,
code that uses ``out=...`` a lot on small arrays should see a nice
difference.

seberg commented:
Just to note, this one should be almost ready, but I should prod it a little bit, and test against astropy (and maybe dask) to make sure they are fine with the `__array_ufunc__` changes (the changes are within the specifications, but if someone breaks those, we may have to help them out a bit).

mhvk commented on numpy/core/src/common/npy_argparse.h:
```diff
@@ -0,0 +1,96 @@
+ *     PyObject *argument1, *argument3;
+ *     int argument2 = -1;
+ *     if (npy_parse_arguments("method", args, len_args, kwnames),
+ *                "argument1", NULL, &argument1,
```
I wondered quite a bit about whether this implementation perhaps should be shared a bit with the `PREPARE` macro, but in the end it does seem that this is quite simple and logical, keeping keyword names, their actions, and their destinations together, etc. So, I guess I ended with a :+1:

seberg replied:
Oh, this should be in now, does it still show up in the diff?  Let me rebase quickly to make it hopefully go away.  (will make a note of the above comments if they disappear though.)

I tried for a while to do that, but I don't think it is possible.  IIRC there was a gcc specific solution (not C99) to be able to give a return value and declare a variable.

mhvk reviewed (commented):
I had a cursory look just to see what was up, and overall it seems a really nice change. I mostly looked at the ufunc code, which is considerably simplified. Great! If it is ready for more detailed review, I can try to make some time (though perhaps just focussing on the ufunc stuff; it is a *lot*, though I can see it is not so easy to separate).

seberg commented:
@mhvk yeah, it would be great to move this along. Because I need to attack the ufunc code soon, and separating argument parsing from the rest of the code is pretty much a prerequisite to refactor more.

It seems I messed up with the other PRs (picking a wrong commit during rebase), the whole `asarray` stuff should be part of a separate PR (technically, I should not need to include that here, so let me try to rebase it away so this is less noisy for you).

seberg commented:
Rebased with only the ufunc related changes.  I will put further fixups into their own commits for clarity (I don't really expect much, but maybe an occasional refcount problem in override-paths or so).

seberg commented on numpy/core/src/umath/override.c:
```diff
@@ -86,367 +79,124 @@ get_array_ufunc_overrides(PyObject *args, PyObject *kwds,
-                            "cannot specify both 'axis' and 'axes'");
+    static PyObject *out_str = NULL;
+    if (out_str == NULL) {
+        out_str = PyUnicode_InternFromString("out");
```
Just my personal slight preference for keeping definitions local (even if they might end up duplicated).

seberg committed:
DOC: Fixup comments based on review and remove TODO

The TypeError TODO is also reflected in the tests I think, the
other TODO seems pretty unnecessary (most users will go through
`tp_vectorcall` anyway, so micro-optimizing that is not all that
relevant).

seberg committed:
DOC: Fixup comment style for now

mhvk commented on numpy/core/src/umath/override.c:
```diff
@@ -86,367 +79,124 @@ get_array_ufunc_overrides(PyObject *args, PyObject *kwds,
+/*
+ * ufunc() and ufunc.outer() accept 'sig' or 'signature';
+ * normalize to 'signature'
+ */
```
Maybe add to the comment that by the time this is called it is no longer possible for both to be present.

mhvk commented on numpy/core/src/umath/override.c:
```diff
@@ -86,367 +79,124 @@ get_array_ufunc_overrides(PyObject *args, PyObject *kwds,
 {
     /*
-     * ufunc.accumulate(a[, axis, dtype, out])
+     * If the keywords include `sig` rename to `signature`. An error
```
Delete "An error ..." as this is done in `ufuncobject.c`

mhvk commented on numpy/core/src/umath/override.c:
```diff
@@ -460,31 +210,26 @@ normalize_at_args(PyUFuncObject *ufunc, PyObject *args,
     /*
      * Check inputs for overrides
      */
     num_override_args = get_array_ufunc_overrides(
```
This change means that one can no longer override on `where` - I think this is OK, but just to be sure.

seberg replied:
Hmmm, that would be a bigger change then I realized.  But I think it is effectively unchanged, because by NumPy rules `where` has to be passed as a keyword argument, and we ever checked `out` when it comes to kwargs?

seberg replied:
Oh, technically you could pass `where` by position to `ufunc.reduce` and get dispatching.  That seems to border on insanity though :) (for one you also have to pass `initial` and everything else.)

Will write a release note now and include this.

mhvk replied:
Indeed, I don't think this will ever have been done in the wild!

mhvk commented on numpy/core/src/umath/override.c:
```diff
@@ -495,104 +240,58 @@ PyUFunc_CheckOverride(PyUFuncObject *ufunc, char *method,
 
     /*
-     * Normalize ufunc arguments.
+     * Normalize ufunc arguments, note that args does not hold any positional
```
This note was very confusing to me: maybe "note that any input and output arrays in args have already been stored in `in_args` and `out_args`"

The comment about `len_args` seems false (at least for reductions).

seberg replied:
Hmmmpf, this probably made a bit more sense in the old (more complicated) iteration.  Adopting your comment.

mhvk commented on numpy/core/src/umath/override.c:
```diff
@@ -495,104 +240,58 @@ PyUFunc_CheckOverride(PyUFuncObject *ufunc, char *method,
+    /*
+     * Reduce-like methods can pass keyword arguments also by position,
+     * in which case the additional positional arguments have to be copied
+     * into the keyword argument dictionary. The __call__ method has to
```
`__call__` and `outer` methods.

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -2637,6 +2273,8 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
 
     /* Possibly remap axes. */
     if (axes != NULL || axis != NULL) {
+        assert((axes == NULL) || (axis == NULL));
```
Stupid and small, but perhaps `assert(!(axes !=  NULL && axis != NULL))`

mhvk reviewed (commented):
@seberg - this is very nice indeed!!!! I think the code is *much* cleaner now. My comments are basically all nitpicks, with perhaps the only one to be thought through being the one about the type of error to be raised when a tuple of outputs has the wrong size.

seberg commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -3274,38 +2850,52 @@ PyUFunc_GenericFunction_int(PyUFuncObject *ufunc,
+        return -1;
+    }
+
+    PyObject *res = ufunc_generic_call_mps(ufunc, args, kwds, op);
```
Renamed it to `_with_operands`, but you are right and I will make a PR after this to just remove the whole thing. Since nobody complained yet, I assume we can go ahead with disabling `PyUFunc_GenericFunction` for good.

seberg commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -4432,32 +4018,101 @@ PyUFunc_Reduceat(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *ind,
+{
+    if (PyTuple_CheckExact(out_obj)) {
+        if (PyTuple_GET_SIZE(out_obj) != nout) {
+            // TODO: Was ValueError (and error order changed)
```
Trying to turn it back into a `ValueError`. I don't remember if I had more reason to change it (will have to see if the tests need adapting later).

seberg commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -4721,37 +4446,324 @@ PyUFunc_GenericReduction(PyUFuncObject *ufunc, PyObject *args,
 
-    errval = PyUFunc_CheckOverride(ufunc, "__call__", args, kwds, &override);
-    if (errval) {
+    /* Deprecated PyUfunc_GenericFunction support only */
```
Yeah, let me also rename `mps` to `ops` or `operands` everywhere.  I honestly don't even know what `mps` is supposed to mean...

seberg commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -3274,38 +2850,52 @@ PyUFunc_GenericFunction_int(PyUFuncObject *ufunc,
 PyUFunc_GenericFunction(PyUFuncObject *ufunc,
         PyObject *args, PyObject *kwds, PyArrayObject **op)
 {
     /* NumPy 1.19, 2020-01-24 */
```
Yes, we can I think. But as you said, lets do it in a followup.  Normal is 2 versions of deprecation, which is now what we got.

seberg commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -5713,6 +5764,15 @@ ufunc_at(PyUFuncObject *ufunc, PyObject *args)
                         "second operand needed for ufunc");
         return NULL;
     }
+    errval = PyUFunc_CheckOverride(ufunc, "at",
```
This seems a bit more symmetric since in all other cases we parse arguments first now? I think I slightly prefer to keep it.

mhvk replied:
OK, that makes sense. Looking again, I see you indeed only get objects out (like elsewhere), so presumably that is very quick.

seberg commented on numpy/core/src/umath/override.c:
```diff
@@ -1,13 +1,15 @@
 #define NPY_NO_DEPRECATED_API NPY_API_VERSION
 #define NO_IMPORT_ARRAY
 
+#include "npy_argparse.h"
```
Good spotting.  This was needed in a very old first version that juggled both `FASTCALL` and the old call protocol.

seberg committed:
MAINT: Address comments by Marten

This fixes the review comments by Marten, mainly small cleanups.
The largest change is that it is now back to a ValueError if the
out tuple has the wrong length.

Co-Authored-By: Marten van Kerkwijk <mhvk@astro.utoronto.ca>

seberg committed:
DOC: Add release notes for ufunc (and FASTCALL) refactor

seberg committed:
BUG: Add missing decref to `dtype` parsing

seberg committed:
BUG: Add missing `static` leading to a (harmless) ref-leak

seberg commented:
OK, leak checker flushed out two small remaining issues (there are more issues unrelated, I had not run `pytest-leaks` in too long...).  (Note: I rebased on `main` the first commits are not actually modified.)

From my side, I consider this finished now.  I don't think the stricter argument parsing should be problematic.

seberg commented:
Hmmmpf, there is a small "conflict" with gh-18670, in the `outer` logic.  It doesn't matter much which one goes in first, but either will need rebasing to ensure a small leak doesn't reappear.

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -849,103 +849,12 @@ ufunc_get_name_cstr(PyUFuncObject *ufunc) {
  */
 NPY_NO_EXPORT int
-_subok_converter(PyObject *obj, int *subok)
+_subok_converter(PyObject *obj, npy_bool *subok)
```
Is it worrying that this is stated not to be covered?

seberg replied:
Weird, it definitely is tested.  Changing the `NPY_NO_EXPORT` to `static` since that seems unnecessary. Maybe that helps, otherwise it is a coverage false-positive.

mhvk reviewed (commented):
Only some nitpicks left, really...  Might be good to get another pair of eyes, just in case.

seberg commented on doc/release/upcoming_changes/15271.performance.rst:
```diff
@@ -0,0 +1,16 @@
+In the following cases NumPy now has a much reduced call
+overhead:
+
+* The array creation functions ``np.asarray``,
```
Yeah, this part is already merged, thought I will make one note for all.  Could probably split things up.

mhvk replied:
It would mostly be for people to be able to trace back from the change-log, so perhaps just referencing the various PRs suffices?

seberg replied:
Added direct links manually and made the file (the link is auto-inserted after the note) point to the "initial" FASTCALL implementing PR.  But I am happy to just split this out or followup in another PR as well, it doesn't have to be added in this PR after all.

The rebase on main (with the conflict) is done.

seberg committed:
MAINT: Declare functions `static` rather than `NPY_NO_EXPORT`

These functions do not require `NPY_NO_EXPORT`. One of them was
(incorrectly) flagged by code coverage. Maybe this helps...

seberg committed:
DOC: Fixup ufunc-argparse release notes and code comments

This is based on Marten's review again.

seberg committed:
BUG: Fix bug(s) introduced due to rebase/merge conflict

charris commented:
If there are two things going on in this PR it might be best to split them.

seberg committed:
DOC: Link all three PRs and use the first PR as the "base" (filename)

seberg commented:
I had not written release notes for gh-15269 and gh-18642, since they should not have any compatibility issues.  But as the PRs do similar things on different part of the code-base I thought it fits together in the release notes.

Other than the release notes thing, this PR is a bit annoyingly large, but that is mostly the nature of the beast, unfortunately.  There are some side battles here, but those won't be easy to split out I think.

mhvk commented on doc/release/upcoming_changes/15271.compatibility.rst:
```diff
@@ -0,0 +1,31 @@
+
+``__array_ufunc__`` and additional positional arguments
+-------------------------------------------------------
+Previously, all positionally passed arguments were check for
```
Sorry, missed a small mistake: check -> checked

seberg replied:
Thanks.  Yeah I have to remember to add `[skip CI]` more often.  Seems it didn't pan out, because I pushed before CI was finished I guess.

mhvk reviewed (approved):
Thanks. To me, this looks all OK!

mhvk commented:
p.s. Minus the last spelling mistake, that is. Note that skipping CI on actions is now possible, by adding [skip CI] to the commit message - see https://github.blog/changelog/2021-02-08-github-actions-skip-pull-request-and-push-workflows-with-skip-ci/

seberg committed:
DOC: Fix spelling in release note

[skip CI]

charris commented:
Thanks Sebastian.
